src/components/cl/hier/allgatherv/unpack.c

```diff
@@ -68,48 +68,89 @@ ucc_status_t ucc_cl_hier_allgatherv_unpack_start(ucc_coll_task_t *task)
     size_t                      dst_rank_disp;
+    ucc_rank_t                 *node_leaders = NULL;
+    ucc_rank_t                 *per_node_leaders = NULL;
+    ucc_sbgp_t                 *all_nodes = NULL;
+    ucc_sbgp_t                 *node_leaders_sbgp = NULL;
+    int                         n_nodes;
+    ucc_rank_t                  node_id;
+    ucc_rank_t                  node_leader_idx;
+    size_t                      src_rank_disp;
 
     UCC_CHECK_GOTO(
         ucc_coll_task_get_executor(&schedule->super, &exec),
         out, status);
     eargs.task_type = UCC_EE_EXECUTOR_TASK_COPY;
 
-    *n_tasks        = 0;
-    src_rank_disp   = 0;
+    *n_tasks = 0;
+
+    // Get the node leaders
+    UCC_CHECK_GOTO(
+        ucc_topo_get_node_leaders(topo, &node_leaders, &per_node_leaders),
+        out, status);
+    
+    // Get the node leaders subgroup for proper rank mapping
+    node_leaders_sbgp = ucc_topo_get_sbgp(topo, UCC_SBGP_NODE_LEADERS);
+    if (!SBGP_EXISTS(cl_team, NODE_LEADERS)) {
+        status = UCC_ERR_NOT_SUPPORTED;
+        goto out;
+    }
+    
+    UCC_CHECK_GOTO(
+        ucc_topo_get_all_nodes(topo, &all_nodes, &n_nodes),
+        out, status);
 
     for (i = 0; i < team_size; i++) {
-        src_rank_count = ucc_coll_args_get_count(args, args->src.info_v.counts,
-                                                 i);
-        dst_rank_count = ucc_coll_args_get_count(args, args->dst.info_v.counts,
-                                                 i);
-        dst_rank_disp  = ucc_coll_args_get_displacement(
-                                args, args->dst.info_v.displacements, i);
-        ucc_assert(src_rank_count * src_dt_size ==
-                   dst_rank_count * dst_dt_size);
-        eargs.copy.src  = PTR_OFFSET(
-                            args->src.info_v.buffer,
-                            src_rank_disp * src_dt_size);
-        eargs.copy.dst  = PTR_OFFSET(
-                            args->dst.info_v.buffer,
-                            dst_rank_disp * dst_dt_size);
-        eargs.copy.len  = dst_rank_count * dst_dt_size;
+        // Get destination count and displacement (in team rank order)
+        dst_rank_count = ucc_coll_args_get_count(args, args->dst.info_v.counts, i);
+        dst_rank_disp = ucc_coll_args_get_displacement(args, args->dst.info_v.displacements, i);
+
+        // Find which node leader this rank belongs to
+        ucc_rank_t leader_team_rank = node_leaders[i];
+        
+        // Find the position of this leader in the node_leaders_sbgp
+        node_leader_idx = ucc_ep_map_local_rank(node_leaders_sbgp->map, leader_team_rank);
+        
+        // Get source displacement from leader_displacements (passed in src.info_v)
+        src_rank_disp = ucc_coll_args_get_displacement(
+            args, args->src.info_v.displacements, node_leader_idx);
+            
+        // Find the correct node for this rank
+        for (node_id = 0; node_id < n_nodes; node_id++) {
```

Comment:
I think this loop can be removed so long as the ranks in the node leader subgroup mapping are in the same order as they are in the team. I've seen this isn't the case though. @Sergei-Lebedev do you think I should try to hunt down the cause and fix it?

Reply:
can you pls elaborate in what case this loop can be removed? you still need to find correct src_rank_disp

Reply:
I removed the loop and added an assert. The loop was just doing what ep_map_local_rank does in the updated version